### PR TITLE
A comprehensive guide for connecting R to Redshift post

### PR DESCRIPTION
**Pull request by sicarul**
A guide through all the available drivers and tools to make your life easier when using Redshift from R and/or RStudio

### CODE REVIEW

**Comment by sicarul** [comment]
Blogpost capture:
screencapture-localhost-4000-blog-a-comprehensive-guide-for-connecting-with-r-to-redshift-1513031019280

**Comment by sicarul** [comment]
We can get something better for the logo, and hosted by us. Is there a guide on how to make a good logo? Should i request it to someone?

**Comment by sicarul** [comment]
I have crafted this simple logo to not bother design, i think it's good enough

image
connectingredshift

**Comment by unicodeveloper** [comment]
Reach out to `@barbu` on slack for the media assets: Logo and Color code.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 1]
```diff
@@ -0,0 +1,262 @@
+---
+layout: post
+title: "A comprehensive guide to connect to Amazon Redshift from R"
+description: A guide through all the available drivers and tools to make your life easier when using Redshift from R and/or RStudio
+date: 2017-12-22 12:00
+category: Technical Guide, Data
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: There should be a third category, `R`.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+dbname="dbname"
+host='my-redshift-url.amazon.com'
+port='5439'
+user='myuser'
+password='mypassword'
+```
+
+## RJDBC
+
+This is the most ["official" way to use Amazon Redshift with R](https://aws.amazon.com/blogs/big-data/connecting-r-with-amazon-redshift/), using the JDBC driver on SQL Workbench is the official way to connect to it according to the documentation, and this driver can be loaded like this:
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: I don't think the `most` is needed here. Please remove.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                 dbname = dbname)
+```
+
+There are two major problems with this driver though: It [lacks SSL support](https://stackoverflow.com/questions/38942118/can-not-connect-postgresqlover-ssl-with-rpostgresql-on-windows), and you don't have transactions like you do with the official driver. In my humble opinion this is the worst choice right now. The good thing is this driver works fine with [dplyr](http://dplyr.tidyverse.org/).
+
+{% include tweet_quote.html quote_text="The RPostgreSQL driver doesn't support SSL-secured Redshift instances" %}
+
+## RPostgres
+
+This is another PostgreSQL library, that has much better support using the libpq library, you'll probably need to install the postgresql development header libraries locally for this one to be installed succesfully. [Check out the official docs for installing instructions if you run into issues.](https://github.com/r-dbi/RPostgres)
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: Please can you reference the libpq library. A link to the GitHub repo or host will do.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+pconn_r <- dbConnect(RPostgres::Postgres(),
+               host = host,
+               port = port,
+               user = user,
+               password = password,
+               dbname = dbname,
+               sslmode='require')
+```
+
+## Gotchas with the libraries
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: `Gotchas with the libraries` => `Gotchas With The Libraries`

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+insert into sicatest values
+('Hello', 1, '2017-10-01', '2017-10-01 20:00:00', TRUE);
+insert into sicatest values
+('Hello ñandu', 9223372036854775807, '2017-12-30', '2017-12-11 23:59:00', FALSE);
+insert into sicatest values
+('こんにちは', 9223372036854775807, '2017-12-30', '2017-12-11 23:59:00', NULL);
+commit;
+```
+
+We are on purpose adding the Ñ letter from spanish, and こんにちは (Kon'nichiwa) which is hello in japanese to have a non-ascii stuff, and see if it's handled correctly.
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: `We are on purpose` => We deliberately added

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+
+
+```r
+dbSendUpdate(jconn, "BEGIN")
+dbSendUpdate(jconn, "COMMENT on table sicatest is 'best table ever'")
+dbSendUpdate(jconn, "COMMIT")
+```
+
+That's why, we can't just rely on one or the other, for maximum efficiency with R + Redshift, we need both drivers, i hope someday we can just rely on one of them, if it's RPostgres or a fork of it, i'll be super happy.
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: `i` => `I` . Please replace the lower case i's with capital letters. `i hope`, `i'll`

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+
+{% include tweet_quote.html quote_text="For maximum efficiency with R + Redshift, we need to use both RJDBC and RPostgreSQL depending on the scenario." %}
+
+Having said that, there's an additional tool which i'd like to introduce to you, which is the `redshiftTools` R Package, it'll add nicely to your toolkit and supports either RJDBC or RPostgres connections. This package is MIT licensed and it's source is available at [https://github.com/sicarul/redshiftTools](https://github.com/sicarul/redshiftTools).
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: `i'd` => `I'd`

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+After installing, you'll have 4 helpful functions to use, which are explained in full detail in [the package's README](https://github.com/sicarul/redshiftTools).
+
+`rs_create_statement`: Generates the SQL statement to create a table based on the structure of a data.frame. It allows you to specify sort key, dist key and if you want to allow compression to be added or not.
+
+`rs_replace_table`: Deletes all records in a table, then uploads the provided data frame into it. It runs as a transaction if provided with an RJDBC connection.
+
+`rs_upsert_table`: Deletes all records matching the provided keys from the uploaded dataset, and then inserts the rows from the dataset. If no keys are provided, it acts as a regular insert.
+
+`rs_create_table`: This just runs rs_create_statement and then rs_replace_table, creating your table and uploading it.
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: rs_create_statement  and rs_replace_table should be highlighted.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+This package is helpful because uploading data with inserts in Redshift is super slow, this is the recommended way of doing replaces and upserts [per the Redshift documentation](http://docs.aws.amazon.com/redshift/latest/dg/t_Loading-data-from-S3.html), which consists in generating various CSV files, uploading them to an S3 bucket and then calling a copy command on the Redshift server, all of that is handled by the package.
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: `which consists in` => `which consists of`

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+                 host = host,
+                 port = port,
+                 user = user,
+                 password = password,
+                 dbname = dbname)
+```
+
+There are two major problems with this driver though: It [lacks SSL support](https://stackoverflow.com/questions/38942118/can-not-connect-postgresqlover-ssl-with-rpostgresql-on-windows), and you don't have transactions like you do with the official driver. In my humble opinion this is the worst choice right now. The good thing is this driver works fine with [dplyr](http://dplyr.tidyverse.org/).
+
+{% include tweet_quote.html quote_text="The RPostgreSQL driver doesn't support SSL-secured Redshift instances" %}
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: Please add a full stop at the end of the sentence.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+related:
+- 2016-12-13-adding-authentication-to-shiny-server
+- 2016-12-06-machine-learning-for-everyone
+- 2017-01-27-machine-learning-for-everyone-part-2-abnormal-behavior
+---
+
+
+[Amazon Redshift](https://aws.amazon.com/redshift/) is one of the hottest databases for Data Warehousing right now, it's one of the most cost-effective solutions available, and allows for integration with many popular BI tools. Unfortunately, the status of the drivers compatibility is a little more shaky, but there is a way to make it work very nicely with [R](https://www.r-project.org/)!
+
+{% include tweet_quote.html quote_text="The status of Redshift drivers compatibility with R is a little shaky, but there is a way to make it work very nicely with R" %}
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: Please add a full stop at the end of the sentence.

File: _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+
+`rs_create_table`: This just runs rs_create_statement and then rs_replace_table, creating your table and uploading it.
+
+This package is helpful because uploading data with inserts in Redshift is super slow, this is the recommended way of doing replaces and upserts [per the Redshift documentation](http://docs.aws.amazon.com/redshift/latest/dg/t_Loading-data-from-S3.html), which consists in generating various CSV files, uploading them to an S3 bucket and then calling a copy command on the Redshift server, all of that is handled by the package.
+
+I hope this guide let's you unlock the full potential of R + Amazon Redshift, two great tools that work very well together when well configured. We use this methodology inside Auth0 and we think it's very useful for other organizations with similar infrastructure.
+
+If you are interested in reading other posts about our work with Data @ Auth0, you may enjoy [Machine Learning for everyone](https://auth0.com/blog/machine-learning-for-everyone/) and [Adding Authentication to Shiny Server in 4 Simple Steps](https://auth0.com/blog/adding-authentication-to-shiny-server/).
```
Comment (unicodeveloper) [inline _posts/2017-12-22-a-comprehensive-guide-for-connecting-with-r-to-redshift.markdown]: Every Blog post on Auth0 needs to have an aside especially when it is not used actively in the post.

So I recommend adding this [aside](https://github.com/auth0/blog/blob/master/_includes/asides/market-basket-links-about.markdown).

All you need to do is add this markdown after this post like so:

`{% include asides/market-baskets-links-about.markdown %}`

**Comment by sicarul** [comment]
Applied your feedback @unicodeveloper ! Thanks

**Pull request merged**